presto-main/src/main/java/io/prestosql/execution/buffer/PagesSerde.java

```diff
@@ -56,86 +56,225 @@ public PagesSerde(BlockEncodingSerde blockEncodingSerde, Optional<Compressor> co
         this.spillCipher = requireNonNull(spillCipher, "spillCipher is null");
     }
 
+    public PagesSerdeContext newContext()
```

Comment:
`PagesSerde` is not thread-safe. It's mean to be used from single-thread. Therefore, we don't need extra `PagesSerdeContext`. We can keep temporary buffers within `PagesSerde` itself.

Reply:
You could do that, and that's what https://github.com/prestodb/presto/pull/13232 added to prestodb for compression buffers. I think there are a couple problems with that approach:
- The memory consumed by these can be non-trivial. In fact, the biggest benefit comes from cases where the serialization buffers are large so then you have to strike a balance with how big you let these buffers become. You also then have to track the memory explicitly since it's no longer "ephemeral"
- Just because `PagesSerde` is not thread safe doesn't mean that it isn't accidentally used from multiple threads on occassion. A quick survey of `PagesSerde` usages around prestodb has some worrying occurrences of `PagesSerde` usages very near (but not inside of) `synchronized` blocks which suggests that they are susceptible to simultaneously using that buffer from multiple threads (eg: [PrestoSparkSerializedPageInput](https://github.com/prestodb/presto/blob/b7ceddfb52ec3d381afd41f18cf19b2d604a6c5e/presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkSerializedPageInput.java#L61)). In my opinion, that's far riskier and more subtle than the this approach of caller managed temporary buffers.
- I suspect that running on machines with non-uniform memory access (NUMA) performance might be worse than this approach in aggregate, even though buffer reuse would be much better. On those machines, the calling thread created `PagesContext` would be allocated within its current core's NUMA node whereas keeping the buffer indefinitely inside of the `PagesSerde` could much more easily result in the thread being scheduled on another core (or socket) than where the buffer was originally allocated and as a result perform very poorly. I haven't attempted to test on a machine with multiple NUMA nodes, so I don't know what the performance is actually like in that scenario.

Reply:
> The memory consumed by these can be non-trivial. In fact, the biggest benefit comes from cases where the serialization buffers are large so then you have to strike a balance with how big you let these buffers become. You also then have to track the memory explicitly since it's no longer "ephemeral"

Do you mean that `PagesSerde` is long lived while context is short lived? If we care about nullyfying buffers, we could add `PagesSerde#reset` method. Having buffers in `PagesSerde` allows to avoid complex aquire/release logic.

> Just because PagesSerde is not thread safe doesn't mean that it isn't accidentally used from multiple threads on occassion.

The reason `PagesSerde` is not thread-safe is because compressors are not thread safe. See `io.airlift.compress.lz4.Lz4Compressor` for example. If `PagesSerde` is used in multi-threaded context I would consider it a bug.

> I suspect that running on machines with non-uniform memory access (NUMA) performance might be worse

This could be solved (if it is a problem) by optinal `PagesSerde#reset` method

Reply:
> Do you mean that PagesSerde is long lived while context is short lived? If we care about nullyfying buffers, we could add PagesSerde#reset method. Having buffers in PagesSerde allows to avoid complex aquire/release logic.

Yes, that's the biggest point here. I'm not sure having a `PagesSerde#reset()` method actually simplifies the situation because:
- it becomes ambiguous about when and why to call `reset()`
- if you let some `PagesSerde` instances be longer-lived, how large do you let those buffers get? Do you need to reduce the max buffer size in those scenarios to avoid consuming too much memory? Do you need to track that memory somewhere?

I agree that adding a try-with-resources block around all of the calls to serialize / deserialize isn't exactly pretty or convenient, but I think it's a better option than allocation all of these buffers on-demand (like happens today) and that it's still *probably* better than `PagesSerde` owning the buffers and having to wire each instance up into memory tracking and memory revoking (ie: `reset()` even though that would in general allow more reuse than this strategy. If the idea is that you would just call `reset()` on every use, then I think this is still the clearer idiom.

Reply:
> if you let some PagesSerde instances be longer-lived, how large do you let those buffers get?

We don't have to make `PagesSerde` instances long lived. Instead of passing `PagesSerde` to operator (like `ExchangeOperator`), we could simply pass `PagesSerdeFactory` or `Supplier<PagesSerde>`. This way `PagesSerde` becomes a `PagesSerdeContext` itself.

Reply:
It seems to me like this would just be shifting the same problem around compared to having a `PagesSerdeContext`. Sure, you could make `PagesSerde` instances relatively ephemeral instead of having a `PagesSerdeContext` but then you have to either re-create the `Compressor` / `Decompressor` each time you create a `PagesSerde` (which would not be particularly efficient since `Lz4Compressor` has that internal 16KiB "table") or keep those around on some other object and pass them in each time along with the `AesSpillCipher` since those keys are also ephemeral and only associated with a single `SpillCipher` instance for it's lifespan. You still have the long-lived stateful entity (in my case, `PagesSerde` in your case `Supplier<PagesSerde>`) and a short lived temporary buffer reusing entity.

Reply:
> if you let some PagesSerde instances be longer-lived, how large do you let those buffers get? Do you need to reduce the max buffer size in those scenarios to avoid consuming too much memory? Do you need to track that memory somewhere?

When we have long-lived objects of larger size, ideally we should account for their memory using `LocalMemoryContext` or `AggregatedMemoryContext`.

> which would not be particularly efficient since Lz4Compressor has that internal 16KiB "table"

That could be a concern (albeit a small one), but serialized page/compression/decompression buffers would be much larger than that (and they would need to be created too if compression is on). Generally, that would be a problem only when a single page is serialized/deserialized (which is the case of `ExchangeOperator` or `TaskOutputOperator`), but we should avoid that anyway.


Additional notes:
* `PagesSerde` is contextual object. No need to have another one.
* `PagesSerdeContext` contains aquire/release logic which won't be needed if `PagesSerde` stored buffers itself.
* `PagesSerdeContext` is `PagesSerde` specific. We would rather have generic `BufferPool` that could be also used in `BlockSerde`